pkg/controllerutil/image_util.go

```diff
@@ -0,0 +1,181 @@
+type RegistryNamespaceConfig struct {
+	From string
+	To   string
+}
+
+type RegistryConfig struct {
+	From       string
+	To         string
+	Namespaces []RegistryNamespaceConfig
+}
+
+type RegistriesConfig struct {
+	DefaultRegistry  string
+	DefaultNamespace string
+	Registries       []RegistryConfig
+}
+
+var registriesConfig = RegistriesConfig{}
+
+func init() {
+	ReloadRegistryConfig()
+}
+
+func ReloadRegistryConfig() {
+	// TODO: this is needed in componnet controller test, is there a better way?
+	registriesConfig = RegistriesConfig{}
+	if err := viper.UnmarshalKey(constant.CfgRegistries, &registriesConfig); err != nil {
+		panic(err)
+	}
+
+	for _, registry := range registriesConfig.Registries {
+		if len(registry.From) == 0 {
+			panic("from can't be empty")
+		}
+
+		for _, namespace := range registry.Namespaces {
+			if len(namespace.From) == 0 || len(namespace.To) == 0 {
+				panic("namespace can't be empty")
+			}
+		}
+	}
+}
+
+// For a detailed explanation of an image's format, see:
+// https://pkg.go.dev/github.com/distribution/reference
+
+// if registry is omitted, the default (docker hub) will be added.
+// if namespace is omiited when using docker hub, the default namespace (library) will be added.
+func parseImageName(image string) (
+	host string, namespace string, repository string, remainder string, err error,
+) {
+	named, err := reference.ParseNormalizedNamed(image)
+	if err != nil {
+		return
+	}
+
+	tagged, ok := named.(reference.Tagged)
+	if ok {
+		remainder += ":" + tagged.Tag()
+	}
+
+	digested, ok := named.(reference.Digested)
+	if ok {
+		remainder += "@" + digested.Digest().String()
+	}
+
+	host = reference.Domain(named)
+
+	pathSplit := strings.Split(reference.Path(named), "/")
+	if len(pathSplit) > 1 {
+		namespace = strings.Join(pathSplit[:len(pathSplit)-1], "/")
+	}
+	repository = pathSplit[len(pathSplit)-1]
+
+	return
+}
+
+func ReplaceImageRegistry(image string) (string, error) {
+	registry, namespace, repository, remainder, err := parseImageName(image)
+	if err != nil {
+		return "", err
+	}
+
+	chooseRegistry := func() string {
+		if registriesConfig.DefaultRegistry != "" {
+			return registriesConfig.DefaultRegistry
+		} else {
+			return registry
+		}
+	}
+
+	chooseNamespace := func() string {
+		if registriesConfig.DefaultNamespace != "" {
+			return registriesConfig.DefaultNamespace
+		} else {
+			return namespace
+		}
+	}
+
+	var dstRegistry, dstNamespace string
+	for _, registryMapping := range registriesConfig.Registries {
+		if registryMapping.From == registry {
+			if len(registryMapping.To) != 0 {
+				dstRegistry = registryMapping.To
+			} else {
+				dstRegistry = chooseRegistry()
+			}
+
+			for _, namespaceConf := range registryMapping.Namespaces {
+				if namespace == namespaceConf.From {
+					dstNamespace = namespaceConf.To
+					break
+				}
+			}
+
+			if dstNamespace == "" {
+				dstNamespace = namespace
```

Comment:
so if namespace is not found, the original namespace rather than the defaultNamespace is used.

Reply:
Maybe add a `DefaultNamespace` field to the `RegistryConfig` struct:

```go
type RegistryConfig struct {
	From       string
	To         string
	DefaultNamespace string
	Namespaces []RegistryNamespaceConfig
}
```

When the namespace is not fount in `Namespaces`, use the DefaultNamespace if it's not empty, otherwise use the original namespace.

Reply:
I think simply use the global defaultNamespace is enough.